perceval/algorithm/sampler.py

```diff
@@ -203,39 +219,80 @@ def _samples_wrapper(self, max_samples: int = None, progress_callback: callable
+            if not self._check_sample_shot_iterator():
+                raise RuntimeError("Local sampling simumation requires max_samples and/or max_shots parameters")
         if max_samples is None:
             max_samples = self.SAMPLES_MAX_COUNT
-        results = {'results_list':[]}
+        self._max_samples = max_samples
+        self._max_shots = max_shots
+        default_it = self._it_default_parameters()
+        results = {'results_list': []}
         for idx, it in enumerate(self._iterator):
-            self._apply_iteration(it)
-            results['results_list'].append(self._processor.samples(max_samples, max_shots))
+            self._processor._simulator = None  # Reset any possible cached parameter
+            self._apply_iteration(default_it | it)
+            results['results_list'].append(self._processor.samples(self._max_samples, self._max_shots))
             results['results_list'][-1]['iteration'] = it
             if progress_callback is not None:
-                progress_callback((idx+1)/len(self._iterator))
+                progress_callback((idx + 1) / len(self._iterator))
+        self._apply_iteration(default_it)  # restore default parameters
         return results
 
     def _apply_iteration(self, it):
-        if 'circuit_params' in it:
+        for key, val in it.items():
+            try:
+                self.__getattribute__(f"_set_{key}")(val)
+            except AttributeError:
+                pass  # Raise NotImplementedError ?
+
+    def _set_circuit_params(self, params: dict):
+        if params:
             circuit_params = self._processor.get_circuit_parameters()
-            for name, value in it['circuit_params'].items():
-                circuit_params[name].set_value(value)
-        if 'input_state' in it:
-            self._processor.with_input(it['input_state'])
-        if 'min_detected_photons' in it:
-            self._processor.min_detected_photons_filter(it['min_detected_photons'])
+            for name, value in params.items():
+                if value is not None:
+                    circuit_params[name].set_value(value)
+
+    def _set_input_state(self, input_state: BasicState):
+        self._processor.with_input(input_state)
+
+    def _set_min_detected_photons(self, count: int):
+        self._processor.min_detected_photons_filter(count)
+        if count is None:
+            self._processor.parameters.pop("min_detected_photons")
+
+    def _set_max_samples(self, val: int):
+        self._max_samples = val
+
+    def _set_max_shots(self, val: int):
+        self._max_shots = val
+
+    def _set_noise(self, noise: NoiseModel):
+        self._processor.noise = noise
+
+    def _it_default_parameters(self) -> dict:
+        """Creates an iteration with default parameters"""
+        return {"circuit_params": {k: v._value for k, v in self._processor.get_circuit_parameters().items()},
```

Comment:
`_value` is None so it will not be set. It has to be set for each iteration, but there is no check that all parameters always have a value before computing